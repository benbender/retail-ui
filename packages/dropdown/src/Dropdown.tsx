import { useControllableValue } from 'ahooks'
import clsx from 'clsx'
import * as React from 'react'

import { DropdownProvider, DropdownValue } from './DropdownContext'
import { DropdownStyles } from './styles'

type ReactDivProps = React.HTMLAttributes<HTMLDivElement>
type Ref = HTMLDivElement

export interface DropdownProps {
  className?: string
  isOpen?: boolean
  onChangeOpen?: (isOpen: boolean) => void
  defaultOpen?: boolean
  children?: ((value: DropdownValue) => React.ReactNode) | React.ReactNode
}

export const Dropdown = React.forwardRef<Ref, ReactDivProps & DropdownProps>(
  (props, ref) => {
    const { children, isOpen: isOpenProp, onChangeOpen, ...rest } = props

    const [isOpen, setIsOpen] = useControllableValue<boolean>(props, {
      defaultValue: false,
      defaultValuePropName: `defaultOpen`,
      valuePropName: `isOpen`,
      trigger: `onChangeOpen`,
    })

    const toggleOpen = React.useCallback(
      (newOpen?: boolean) => {
        setIsOpen(newOpen ?? !isOpen)
      },
      [isOpen],
    )

    const dropdownValue = React.useMemo(
      () => ({
        isOpen: Boolean(isOpen),
        toggleOpen,
      }),
      [isOpen, toggleOpen],
    )

    const cls = clsx(DropdownStyles.base)

    return (
      <DropdownProvider value={dropdownValue}>
        <div className={cls} ref={ref} {...rest}>
          {children}
        </div>
      </DropdownProvider>
    )
  },
)

Dropdown.displayName = 'Dropdown'
