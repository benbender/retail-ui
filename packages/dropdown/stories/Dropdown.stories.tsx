import { Button } from '@retail-ui/button'
import { ChevronDownOutline } from '@retail-ui/heroicons'
import * as React from 'react'

import {
  Dropdown,
  DropdownButton,
  DropdownItem,
  DropdownList,
  DropdownProps,
} from '../src'

const meta = {
  title: 'Retail-UI/Dropdown',
  component: Dropdown,
}

export default meta

const Template = (args: DropdownProps) => {
  const [isOpen, setIsOpen] = React.useState(false)

  return (
    <Dropdown
      {...args}
      data-testid={meta.title}
      isOpen={isOpen}
      onChangeOpen={(state) => setIsOpen(state)}
    >
      <Button onClick={() => setIsOpen((state) => !state)}>
        Control dropdown
      </Button>
      <DropdownButton
        className="relative align-middle rounded-md focus:outline-none focus:shadow-outline-purple"
        aria-label="Notifications"
        aria-haspopup="true"
        onClick={() => {
          console.log('something')
        }}
      >
        {!isOpen && <ChevronDownOutline />}
        {isOpen && `Up`}
      </DropdownButton>
      <DropdownList>
        <DropdownItem className="justify-between">
          <span>Messages</span>
          <span className="inline-flex items-center justify-center px-2 py-1 text-xs font-bold leading-none text-red-600 bg-red-100 rounded-full dark:text-red-100 dark:bg-red-600">
            13
          </span>
        </DropdownItem>
        <DropdownItem className="justify-between">
          <span>Sales</span>
          <span className="inline-flex items-center justify-center px-2 py-1 text-xs font-bold leading-none text-red-600 bg-red-100 rounded-full dark:text-red-100 dark:bg-red-600">
            2
          </span>
        </DropdownItem>
        <DropdownItem className="justify-between">
          <span>Alerts</span>
        </DropdownItem>
      </DropdownList>
    </Dropdown>
  )
}

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Dropdown',
}
