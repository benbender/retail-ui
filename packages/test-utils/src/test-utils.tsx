import { defaultTheme, ThemeProvider } from '@retail-ui/theme'
import {
  render as RtlRender,
  RenderOptions,
  RenderResult,
} from '@testing-library/react'
import * as React from 'react'

type UI = Parameters<typeof RtlRender>[0]

// UI-less passthrough fallback to prevent using conditional logic in render
function ChildrenPassthrough({ children }: { children: React.ReactElement }) {
  return children
}

export interface TestOptions extends Omit<RenderOptions, 'wrapper'> {
  /**
   * optional additional wrapper, e.g. Context
   *
   * @example
   * ```ts
   * // single wrapper
   * render(<MyConponent />, {
   *  wrapper: MyContext
   * });
   *
   * // multiple wrapper
   * render(<MyConponent />, {
   *  wrapper: ({ children }) => (
   *    <ContextA>
   *      <ContextB>
   *        {children}
   *      <ContextB />
   *    <ContextA />
   *  )
   * });
   *
   * ```
   */
  wrapper?: typeof ChildrenPassthrough
}

/**
 * Custom render for @testing-library/react
 *
 * @see https://testing-library.com/docs/react-testing-library/setup#custom-render
 * @param component the component under test
 * @param options customized test options
 */
export const render = (
  ui: UI,
  { wrapper: Wrapper = ChildrenPassthrough, ...options }: TestOptions = {},
): RenderResult => {
  return RtlRender(
    <ThemeProvider value={{ theme: defaultTheme }}>
      <Wrapper>{ui}</Wrapper>
    </ThemeProvider>,
    options,
  )
}

export { fireEvent } from '@testing-library/react'
