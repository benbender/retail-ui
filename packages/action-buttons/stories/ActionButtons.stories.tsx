import * as React from 'react'

import { ActionButtons, ActionButtonsProps } from '../src'

const meta = {
  title: 'Retail-UI/ActionButtons',
  component: ActionButtons,
}

export default meta

const Template = (args: ActionButtonsProps) => (
  <ActionButtons {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {}
