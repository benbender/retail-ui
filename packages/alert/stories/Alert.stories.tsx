import * as React from 'react'

import { Alert, AlertProps } from '../src'

const meta = {
  title: 'Retail-UI/Alert',
  component: Alert,
}

export default meta

const Template = (args: AlertProps) => (
  <Alert {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'A alert that informs you of stuff',
}
