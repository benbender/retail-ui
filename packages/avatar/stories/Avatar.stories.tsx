import * as React from 'react'

import { Avatar, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/Avatar',
  component: Avatar,
}

export default meta

const Template = (args: AvatarProps) => (
  <Avatar {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  src: `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`,
}
