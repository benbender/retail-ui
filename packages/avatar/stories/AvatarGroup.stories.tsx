import * as React from 'react'

import { Avatar, AvatarGroup, AvatarProps } from '../src'

const meta = {
  title: 'Retail-UI/AvatarGroup',
  component: AvatarGroup,
}

export default meta

const src = `https://pbs.twimg.com/profile_images/1285102166709919744/XvlWrJny_400x400.jpg`

const Group = (args: AvatarProps) => (
  <AvatarGroup data-testid={meta.title}>
    <Avatar src={src} {...args} />
    <Avatar title="PJ" size="base" {...args} />
    {/* <Avatar icon={UserSolid} {...args} /> */}
  </AvatarGroup>
)

export const Simple = Group.bind({})
// @ts-ignore
Simple.args = {}
