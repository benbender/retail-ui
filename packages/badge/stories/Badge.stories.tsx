import * as React from 'react'

import { Badge, BadgeProps } from '../src'

const meta = {
  title: 'Retail-UI/Badge',
  component: Badge,
}

export default meta

const Template = (args: BadgeProps) => (
  <Badge {...args} data-testid={meta.title} />
)

export const Simple = Template.bind({})
// @ts-ignore
Simple.args = {
  children: 'Badge',
}
