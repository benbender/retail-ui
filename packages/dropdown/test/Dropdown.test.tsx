import { render } from '@retail-ui/test-utils'
import * as React from 'react'

import * as Dropdown from '../stories/Dropdown.stories'

describe('Dropdown', () => {
  it('renders correctly', () => {
    const { asFragment } = render(<Dropdown.Simple />)
    expect(asFragment()).toMatchSnapshot()
  })
})
