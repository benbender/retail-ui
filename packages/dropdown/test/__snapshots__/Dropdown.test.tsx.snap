// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown renders correctly 1`] = `
<DocumentFragment>
  <div
    class="relative inline-block"
    data-testid="Retail-UI/Dropdown"
  >
    <button
      class="inline-flex items-center justify-center leading-5 align-bottom transition-colors duration-150 ease-in transform scale-100 border cursor-pointer focus:outline-none dark:focus:shadow-outline-gray active:scale-95 overflow-hidden no-underline select-none text-white bg-purple-600 border-transparent active:shadow-outline-purple focus:bg-purple-700 hover:bg-purple-700 py-2 px-4 text-base font-medium rounded-lg"
    >
      <span
        class="opacity-100"
      >
        Control dropdown
      </span>
    </button>
    <button
      aria-haspopup="true"
      aria-label="Notifications"
      class="relative align-middle rounded-md focus:outline-none focus:shadow-outline-purple"
    >
      <svg
        aria-hidden="true"
        class="w-7 h-7 text-current stroke-current"
        fill="none"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 9l-7 7-7-7"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;
